### ISSUE TITLE
About NNSVR

### ISSUE THREAD

**Issue by caiquanyou** [Klopfe/LSVR#1]
hello，I have read about your paper :LINEAR SUPPORT VECTOR REGRESSION WITH LINEAR CONSTRAINTS; and got some question abouy NNSVR?
Since used SVR for deconvolution, they usually set the negative value to 0 and that's done,what is the different if we use NNSVR to solve the same problem?

**Comment by Klopfe** [comment]
Hi thanks for your interest in our work, 
If you solve SVR first and then set the negative coefficients to zeros, it can lead mathematically to a solution that is far from the solution of the constrained optimization problem including the non negative constraints directly in the optimization problem. 
In the first case when you do SVR and then set the coefficients to zero, you solve the optimization problem for the classical SVR and then you project this results on your non negative constraints. This leads to a solution that is generally not the same solution as the NNSVR.
What we do is that we directly include the constraints in the optimization problem which generally leads to better results as stated in the article.

Feel free to ask more questions if needed,

**Comment by caiquanyou** [comment]
Thanks for your reply and if I&nbsp; want to slove the problem like CIBERSORT you mention in your paper ? Does NNSVR help a lot ? Or can you provide a tutorial on single cell deconvolution?
Thanks again&nbsp; for answering the question!
Best
CQY




&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Klopfe/LSVR"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月6日(星期二) 晚上8:34
收件人:&nbsp;"Klopfe/LSVR"<LSVR@noreply.github.com&gt;;
抄送:&nbsp;"xianmao"<951463554@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [Klopfe/LSVR] About NNSVR (#1)





 
Hi thanks for your interest in our work,
 If you solve SVR first and then set the negative coefficients to zeros, it can lead mathematically to a solution that is far from the solution of the constrained optimization problem including the non negative constraints directly in the optimization problem.
 In the first case when you do SVR and then set the coefficients to zero, you solve the optimization problem for the classical SVR and then you project this results on your non negative constraints. This leads to a solution that is generally not the same solution as the NNSVR.
 What we do is that we directly include the constraints in the optimization problem which generally leads to better results as stated in the article.
 
Feel free to ask more questions if needed,

**Comment by caiquanyou** [comment]
I test the NNSVR using  the single cell and bulk RNA  data ,I  can not figure out the meaning of the outputs? 
Which one should I carry on if I want to calculate the proportion of cell type?
Best
CQY

**Comment by Klopfe** [comment]
Hi, the API we use is the same as scikit-learn. The coefficients are accesible once you have fitted the model by calling your variable containing your model let's call it model for now. 
So you would do something like 
model.fit(X, y)
my_proportions = model.coef_

I hope it helps, the code is not super updates do not hesitate if you have questions/problems

**Comment by caiquanyou** [comment]
I calculate the coef using NNSVR,but still get some negative values?anything wrong ?
&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Klopfe/LSVR"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月12日(星期一) 下午5:32
收件人:&nbsp;"Klopfe/LSVR"<LSVR@noreply.github.com&gt;;
抄送:&nbsp;"xianmao"<951463554@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [Klopfe/LSVR] About NNSVR (#1)





 
Hi, the API we use is the same as scikit-learn. The coefficients are accesible once you have fitted the model by calling your variable containing your model let's call it model for now.
 So you would do something like
 model.fit(X, y)
 my_proportions = model.coef_
 
I hope it helps, the code is not super updates do not hesitate if you have questions/problems

**Comment by Klopfe** [comment]
Hi, What is the size of your data ? How many iterations did the algorithms make ? 
If you set the tolerance to 1e-3 for example, it is possible to have negative values -0.00x because of numerical approximation.

**Comment by Klopfe** [comment]
if it helps you i can provide an easy example with comments for you to try my algorithm. Just let me know

**Comment by caiquanyou** [comment]
&nbsp;If you could provide an easy example,that would be great! Thanks!
&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Klopfe/LSVR"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月13日(星期二) 下午3:56
收件人:&nbsp;"Klopfe/LSVR"<LSVR@noreply.github.com&gt;;
抄送:&nbsp;"xianmao"<951463554@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [Klopfe/LSVR] About NNSVR (#1)





 
if it helps you i can provide an easy example with comments for you to try my algorithm. Just let me know

**Comment by caiquanyou** [comment]
my data size is X:10000x913;y:10000x1
other parameters are all set to default




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Klopfe/LSVR"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月13日(星期二) 下午3:43
收件人:&nbsp;"Klopfe/LSVR"<LSVR@noreply.github.com&gt;;
抄送:&nbsp;"xianmao"<951463554@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [Klopfe/LSVR] About NNSVR (#1)





 
Hi, What is the size of your data ? How many iterations did the algorithms make ?
 If you set the tolerance to 1e-3 for example, it is possible to have negative values -0.00x because of numerical approximation.

**Comment by caiquanyou** [comment]
Oh ,and  the SMO algorithm reached maximal iteration,so does it mean model is not convergent?

**Comment by Klopfe** [comment]
Yes that is why you have negative coefficients. you can increase the number of iterations.

**Comment by caiquanyou** [comment]
It seems take very long time to run.
I think you could provide an simple easy example with comments to try algorithm and that would help me a lot !

Best

**Comment by caiquanyou** [comment]
I try to increase max_iter number but still  have negative coefficients and model is not convergent, any other parameters I  can try to adjust ?
